Helpers/CommonMenuItemPartMigrator.cs

```diff
@@ -0,0 +1,128 @@
+using Etch.OrchardCore.Widgets.Models;
+using GraphQL;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.Autoroute.Models;
+using OrchardCore.ContentFields.Fields;
+using OrchardCore.ContentManagement;
+using OrchardCore.ContentManagement.Records;
+using OrchardCore.ContentManagement.Routing;
+using OrchardCore.Menu.Models;
+using System;
+using System.Threading.Tasks;
+using YesSql;
+
+namespace Etch.OrchardCore.Menu.Helpers
+{
+    public class CommonMenuItemPartMigrator : ICommonMenuItemPartMigrator
+    {
+        private readonly IContentManager _contentManager;
+        private readonly YesSql.ISession _session;
+        private readonly AutorouteOptions _autorouteOptions;
+        private readonly IUrlHelperFactory _urlHelperFactory;
+        private readonly IActionContextAccessor _actionContextAccessor;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly ILogger<CommonMenuItemPartMigrator> _logger;
+
+        public CommonMenuItemPartMigrator(IContentManager contentManager,
+            YesSql.ISession session,
+            IOptions<AutorouteOptions> autorouteOptions,
+            IUrlHelperFactory urlHelperFactory,
+            IActionContextAccessor actionContextAccessor,
+            IHttpContextAccessor httpContextAccessor,
+            ILogger<CommonMenuItemPartMigrator> logger)
+        {
+            _contentManager = contentManager;
+            _session = session;
+            _autorouteOptions = autorouteOptions.Value;
+            _urlHelperFactory = urlHelperFactory;
+            _actionContextAccessor = actionContextAccessor;
+            _httpContextAccessor = httpContextAccessor;
+            _logger = logger;
+        }
+
+        public async Task GetCommonMenuitemDataAsync()
+        {
+            var contentItems = await _session.Query<ContentItem>()
+                .With<ContentItemIndex>()
+                    .Where(x => x.Latest && x.ContentType == "Menu")
+                .ListAsync();
+
+            foreach (var contentItem in contentItems)
+            {
+                var menuItemsListPart = contentItem.As<MenuItemsListPart>();
+
+                foreach (var linkMenuItem in menuItemsListPart.MenuItems)
+                {
+                    var linkDestinationPart = new LinkDestinationPart();
+
+                    try
+                    {
+                        if (linkMenuItem.As<CommonMenuItem>().ExternalUrl.Text != null)
+                        {
+                            linkDestinationPart = new LinkDestinationPart { DestinationUrl = linkMenuItem.As<CommonMenuItem>().ExternalUrl }; ;
+                        }
+                        else
+                        {
+                            var host = _httpContextAccessor.HttpContext.Request.Host.Value;

```

Comment:
What's going on here? We have had discussions offline about relative/absolute URLs and the fact it **must** be a relative URL otherwise moving content becomes a nightmare. Change this to be relative.

Reply:
The issue is here, when I use httpcontextaccessor. it returns to me a path such as "/studio/admin" i assume the admin is there as the migration is running in the background therefore it is returning the "admin" 

so my thinking was to remove admin from the url which would leave me with "/studio/" of which i append the contentitemId autoroute path "homepage" (in this example) which returns a url of "/studio/homepage.

Reply:
You shouldn't need to use the `HttpContext`.

Reply:
what should i be using instead?

Reply:
If you're trying to get the request URL prefix then you should be getting it from the [Shell Settings](https://github.com/OrchardCMS/OrchardCore/blob/5e37e29c54d44bffeeae529f8fd2cf3359734911/src/OrchardCore/OrchardCore.Abstractions/Shell/ShellSettings.cs#L61)